_posts/2023-09-27-onstove-workflow.md

```diff
@@ -0,0 +1,261 @@
+``SCENARIOS``. Then we create another list with all the ``COUNTRIES`` of the study area we want 
+to run, and any other required parameter as ``RESTRICTION``. Then the first rule we want to 
+create is the "final" rule that we call ``all``. This rule will tell Snakemake what is the 
+expected output of the entire workflow, and we define that output as an ``input`` to the rule. In 
+this case, it is a final ``results.pkl`` file for Africa for each scenario in ``SCENARIOS``.
+
+```python
+import os
+cwd = os.getcwd()
+
+SCENARIOS = ['Social_private_benefits', 'Private_benefits']
+RESTRICTION = ['Positive_Benefits']
+
+COUNTRIES = ['AGO', 'BDI', 'BEN', 'BFA', 'BWA', 'CAF', 'CIV', 'CMR',
+             'COD', 'COG', 'DJI', 'ERI', 'ETH', 'GAB', 'GHA', 'GIN',
+             'GMB', 'GNB', 'GNQ', 'KEN', 'LBR', 'LSO', 'MDG', 'MLI',
+             'MOZ', 'MRT', 'MWI', 'NAM', 'NER', 'NGA', 'RWA', 'SDN',
+             'SEN', 'SLE', 'SOM', 'SSD', 'SWZ', 'TCD', 'TGO', 'TZA',
+             'UGA', 'ZAF', 'ZMB', 'ZWE']
+
+rule all:
+    input:
+        expand("Results/Africa/{scenario}/{restriction}/results.pkl",
+               scenario=SCENARIOS,
+               restriction=RESTRICTION)
+```
+
+Then we need to create all intermediate rules required to go from raw data processing until 
+generating the desired final output files. 
+
+### The `process_data` rule <a name="process_data"></a>
+The first rule we need is a ``process_data`` one, 
+where we read an process a set of geospatial datasets. All this datasets should already be 
+downloaded into the system and organized in the specified folder structure. In the 
+[documentation](https://onstove-documentation.readthedocs.io/en/latest/?badge=latest) of 
+the model we provide links to the raw data sources and information on how to use them. 
+Each dataset is then specified as an individual input for the rule. 
+
+Note that, somethimes we use 
+[wildcards](https://snakemake.readthedocs.io/en/stable/snakefiles/rules.html) 
+when defining the path of a dataset. This is done to generalize one rule to different 
+input datasets. Take the case of the `population` input in the `process_data` rule for example, 
+it contains the wildcard `{country}` in its path. From this, Snakemake can then infer which are 
+the countries that need to be run based on the amount of `population` datasets found in the 
+`Population` data folder. 
+
+After the ``input`` we define any parameter needed for the analysis 
+in the ``params`` section. These are parameters that can be used within the defined 
+``script``.  Next, we define the expected ``output`` files after the rule was run. 
+Snakemake will use these information to check that the rule was run successfully and 
+track that the output files are up-to-date. Finally, we specify in the ``script`` section, 
+a python script that should be run in order to perform the desired data processing and 
+processing the ``output`` files.
+
+```python
+rule process_data:
```

Comment:
Could you break this rule into multiple smaller rules?  It may be computationally advantageous if the resulting parallelization was a benefit.

Reply:
Yes this is true, I did consider doing it like that when creating the workflow, but since we were running for 44 countries and processing big spatial datasets, we were constrained by the available memory RAM rather than the cores. So splitting the rule into multiple simpler processes would not have created a considerable benefit. This, however, makes a lot of sense if running a country case for example.